python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -0,0 +1,243 @@
+
+
+import numpy as np
+
+from lsst.daf.butler import Butler, Config, DatasetType
+
+
+def makeTestRepo(root, dataIds, *, config=None, **kwargs):
+    """Create an empty repository with dummy data IDs.
+
+    Parameters
+    ----------
+    root : `str`
+        The location of the root directory for the repository.
+    dataIds : `~collections.abc.Mapping` [`str`, `iterable`]
+        A mapping keyed by the dimensions used in the test. Each value
+        is an iterable of names for that dimension (e.g., detector IDs for
+        `"detector"`). Related dimensions (e.g., instruments and detectors)
+        are linked arbitrarily.
+    config : `lsst.daf.butler.Config`, optional
+        A configuration for the repository (for details, see
+        `lsst.daf.butler.Butler.makeRepo`). If omitted, creates a repository
+        with default dataset and storage types, but optimized for speed.
+    **kwargs
+        Extra arguments to `lsst.daf.butler.Butler.makeRepo`.
+
+    Returns
+    -------
+    butler : `lsst.daf.butler.Butler`
+        A Butler referring to the new repository. This Butler is provided only
+        for additional setup; to keep test cases isolated, it is highly
+        recommended that each test create its own Butler with a
+        unique run/collection. See `makeTestCollection`.
+
+    Notes
+    -----
+    Since the values in ``dataIds`` uniquely determine the repository's
+    data IDs, the fully linked IDs can be recovered by calling
+    `expandUniqueId`, so long as no other code has inserted dimensions into
+    the repository registry.
+    """
+    if not config:
+        config = Config()
+        config[".datastore.cls"] = "lsst.daf.butler.datastores.inMemoryDatastore.InMemoryDatastore"
+        config[".datastore.checksum"] = False  # In case of future changes
+        config[".registry.db"] = "sqlite:///:memory:"
+    Butler.makeRepo(root, config=config, **kwargs)
+    butler = Butler(root, writeable=True)
+    dimensionRecords = _makeRecords(dataIds, butler.registry.dimensions)
+    for dimension, records in dimensionRecords.items():
+        butler.registry.insertDimensionData(dimension, *records)
+    return butler
+
+
+def makeTestCollection(repo):
+    """Create a read/write Butler to a fresh collection.
+
+    Parameters
+    ----------
+    repo : `lsst.daf.butler.Butler`
+        A previously existing Butler to a repository, such as that returned by
+        `~lsst.daf.butler.Butler.makeRepo` or `makeTestRepo`.
+
+    Returns
+    -------
+    butler : `lsst.daf.butler.Butler`
+        A Butler referring to a new collection in the repository at ``root``.
+        The collection is (almost) guaranteed to be new.
+    """
+    # Create a "random" collection name
+    # Speed matters more than cryptographic guarantees
+    collection = "test" + "".join((str(i) for i in np.random.randint(0, 10, size=8)))
+    return Butler(butler=repo, run=collection)
+
+
+def _makeRecords(dataIds, universe):
```

Comment:
The first one isn't; it really does feel like a test helper.  The second one might be more generally useful, but I think it's a special case of more general functionality that we want to make more accessible than it is currently.  So I think I'd file it under a more general work to make `queryDimensions` more convenient rather than moving this function now.